### PR TITLE
[Aichat] Implement OpenAI GPT 4o mini for safety check

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR adds the Open AI GPT 4o model option for the safety layer on `aichat` levels.
For now, we use the `'simple'` moderation system prompt, but will most likely iterate on the prompt soon.
Thus, I add a DCDO-enabled function `get_openai_system_prompt` that returns a prompt based on the key name (default is currently `'simple'`.

For now AWS comprehend is enabled, but we can enable OpenAI and disable AWS comprehend via DCDO.

In `openai_chat_helper`, I opted to create a new function `request_safety_check` since the purpose of this function is to not request chat completion, but to get feedback about safety on a particular text. Thus, there are two parameters, `text` to be checked for `safety` (no conversation history included), and the moderation `system_prompt`. These are the only two 'messages' included in the OpenAI API request. Other differences between this function and `request_chat_completion` is the requested GPT model, API key, and temperature (in case AI Tutor decides to update their temperature used in `request_chat_completion`).

I also created an `AichatSystemPromptHelper` similar to `AitutorSystemPromptHelper` so that we can add/update moderation system prompts to test within this file.

In `openai_safety_check`, I added 2 tries as suggested by @snickell who create a tester branch when setting distinct OpenAI api keys for `aichat`. https://github.com/code-dot-org/code-dot-org/pull/61259/files
Thanks Seth for setting up the keys in AWS secrets for us! 

I hard-coded an invalid response  from `openai_safety_check` - the user is displayed a general error message in the workspace and prompted to try again. And the user chat message is logged in `AichatRequests` with error status.

Screenshot 2024-09-24 at 3 21 00 PM

## Links
[jira](https://codedotorg.atlassian.net/browse/LABS-1039)

## Testing story
- I enabled openai and disabled aws comprehend via dcdo.
- Added a print statement for `details` in `aichat_safety_helper`.
```
      if openai_enabled?(role)
        details = openai_safety_check(text)
        puts "details #{details}"
        return {text: text, blocked_by: 'openai', details: details} if details
      end
```
- I tested with safe messages and message that included profanity and got expected results in responses and `details`.
- I tested adding profane messages to system prompt, retrievals, and both. Got expected results via notifications in chat workspace and `details`.

Screenshot 2024-09-23 at 12 09 55 PM

## Follow-up work
[Engineering Tasks for gpt-4o-mini moderation](https://docs.google.com/document/d/1VVupAlH8HekEiLal6MozW_iVpIxgJ4iwpu9BeDAxyms/edit)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/helpers/aichat_system_prompt_helper.rb
Comment (fisher-alice) [inline dashboard/app/helpers/aichat_system_prompt_helper.rb]: ~~Can return json - definitely open to feedback/ideas on what to include other than flagged text.~~

Comment (fisher-alice) [reply dashboard/app/helpers/aichat_system_prompt_helper.rb]: Actually, I think we want to currently implement the same moderation system prompt used while @snickell was testing. Just fixed a couple typos.

File: dashboard/app/helpers/aichat_safety_helper.rb
Comment (fisher-alice) [inline dashboard/app/helpers/aichat_safety_helper.rb]: Setting the default for openai 'off' for now so @samantha-code can finish testing levels with aws comprehend and webpurify.
Note that on `production` and `levelbuilder` webpurify's DCDO updated to `true`.

File: config/development.yml.erb
Comment (snickell) [inline config/development.yml.erb:68]: Why is this not marked as `!Secret`? That way anyone can try this in dev without having to get a special dev key for their local usage.

Also, the same should added to adhoc, see: https://github.com/code-dot-org/code-dot-org/pull/61259/files

File: dashboard/app/helpers/aichat_safety_helper.rb
Comment (snickell) [reply dashboard/app/helpers/aichat_safety_helper.rb]: I'd suggest we commit with the defaults set how we anticipate going forward (with openai at this point), and set DCDO flags to override to let Sam keep working.

File: dashboard/app/helpers/aichat_safety_helper.rb
Comment (snickell) [inline dashboard/app/helpers/aichat_safety_helper.rb]: I'd suggest simplifying here and baking in the prompt, and changing it when it needs changing. Having "dead code paths" doesn't seem like it improves things, and adds confusion in the future. The only semi-universal software engineering principle I know of is: when given the option, choose the simpler path.

Its probably actually faster and easier to updating a hardcoded string prompt than having to read the code and figure out how to add a key, and its equally as hardcoded either way.

File: dashboard/app/helpers/openai_chat_helper.rb
Comment (snickell) [inline dashboard/app/helpers/openai_chat_helper.rb]: This should be one to match our testing behavior, but I'd suggest not hardcoding a temperature and leaving this param blank (its not required, and this allows openai to optimize for what they think the best default temperature is).

**Review by snickell — CHANGES_REQUESTED** [review]

File: dashboard/app/helpers/openai_chat_helper.rb
Comment (fisher-alice) [reply dashboard/app/helpers/openai_chat_helper.rb]: Oh interesting - thanks!
So just confirming that during your testing you left this param blank?

File: dashboard/app/helpers/aichat_safety_helper.rb
Comment (fisher-alice) [reply dashboard/app/helpers/aichat_safety_helper.rb]: Sure - this makes sense. When I initially wrote this, I wasn't sure if we were going to implement your prompt from testing or the few-shot prompt, but I understand now we'll go with the tested prompt first. Thanks!

File: dashboard/app/helpers/aichat_safety_helper.rb
Comment (fisher-alice) [reply dashboard/app/helpers/aichat_safety_helper.rb]: We'd have to carefully monitor when DTP happens to set flags back and also make sure `production` and `levelbuilder` are in sync. So I think I prefer keeping as is if you're okay with that. Thanks! https://codedotorg.slack.com/archives/C07LQF3AGMA/p1727188797741739

Comment (fisher-alice) [reply dashboard/app/helpers/aichat_safety_helper.rb]: After discussion offline, we'll wait to merge this PR until after @samantha-code is done testing on `production` so setting default to OpenAI. Thanks!

File: dashboard/app/helpers/aichat_safety_helper.rb
Comment (sanchitmalhotra126) [inline dashboard/app/helpers/aichat_safety_helper.rb:39]: For my understanding, does `Retryable` automatically retry if an error was raised in the block (i.e. we don't need to `rescue` it ourselves, right)?

File: dashboard/test/helpers/aichat_safety_helper_test.rb
Comment (sanchitmalhotra126) [inline dashboard/test/helpers/aichat_safety_helper_test.rb]: minor: is this needed? I think this test is just checking user output

File: dashboard/test/helpers/aichat_safety_helper_test.rb
Comment (sanchitmalhotra126) [inline dashboard/test/helpers/aichat_safety_helper_test.rb:1]: Could we add a test to verify the retry logic?

File: dashboard/app/helpers/aichat_safety_helper.rb
Comment (sanchitmalhotra126) [inline dashboard/app/helpers/aichat_safety_helper.rb]: Do we need to return the full text here again? We already include it as part of the full safety response

File: dashboard/app/helpers/aichat_safety_helper.rb
Comment (sanchitmalhotra126) [inline dashboard/app/helpers/aichat_safety_helper.rb:29]: Should the openai check be first (or at least ahead of comprehend)?

File: dashboard/app/helpers/openai_chat_helper.rb
Comment (snickell) [reply dashboard/app/helpers/openai_chat_helper.rb]: Confirmed, I used the default value for this param (which I believe results in a temperature of 1, but I would suggest skipping the param to match test results as closely as possible)

**Review by snickell — APPROVED** [review]

File: dashboard/app/helpers/aichat_safety_helper.rb
Comment (snickell) [reply dashboard/app/helpers/aichat_safety_helper.rb:39]: yes

File: dashboard/test/helpers/aichat_safety_helper_test.rb
Comment (snickell) [reply dashboard/test/helpers/aichat_safety_helper_test.rb:1]: I'd specifically say: "test that it retries if a stubbed `request_safety_check` returns a response.body other than `INAPPROPRIATE` or `OK`."

File: dashboard/app/helpers/aichat_safety_helper.rb
Comment (snickell) [reply dashboard/app/helpers/aichat_safety_helper.rb:29]: From a modeling perspective, I can't see wanting to use a mix of openai and comprehend, so I'm not sure it matters which is first. Arguably it should even be an error to have them both enabled, though I'm not sure its worth enforcing that.

File: dashboard/test/helpers/aichat_safety_helper_test.rb
Comment (sanchitmalhotra126) [reply dashboard/test/helpers/aichat_safety_helper_test.rb:1]: Yep, and also if `request_safety_check` raises an error (right)?

Comment (fisher-alice) [reply dashboard/test/helpers/aichat_safety_helper_test.rb:1]: Sure - just added!

File: dashboard/app/helpers/aichat_safety_helper.rb
Comment (fisher-alice) [reply dashboard/app/helpers/aichat_safety_helper.rb:39]: Added tests to confirm this.

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM!

File: dashboard/app/helpers/openai_chat_helper.rb
Comment (fisher-alice) [reply dashboard/app/helpers/openai_chat_helper.rb]: Link to comment about this temperature setting: https://docs.google.com/document/d/1LB7Pc-X5RU52dUTqN4M8cruDrXodR85Xvo05aBdwAus/edit?disco=AAABWlU2K8Q

**Pull request merged**
